src/ripple/app/tests/Offer.test.cpp

```diff
@@ -235,12 +253,393 @@ class Offer_test : public beast::unit_test::suite
             env.require (balance (carol, USD2 (50)));
         }
     }
+
+    void
+    testInsufficientReserve ()
+    {
+        testcase ("Insufficient Reserve");
+
+        // If an account places an offer and its balance
+        // *before* the transaction began isn't high enough
+        // to meet the reserve *after* the transaction runs,
+        // then no offer should go on the books but if the
+        // offer partially or fully crossed the tx succeeds.
+
+        using namespace jtx;
+
+        auto const gw = Account("gateway");
+        auto const USD = gw["USD"];
+
+        auto const usdOffer = USD(1000);
+        auto const xrpOffer = XRP(1000);
+
+        // No crossing:
+        {
+            Env env(*this);
+            env.fund (XRP(1000000), gw);
+
+            // The fee that's charged for transactions
+            auto const f = env.current ()->fees ().base;
+            auto const r = reserve (env, 0);
+
+            env.fund (r + f, "alice");
+
+            env (
+                trust ("alice", usdOffer), fee(f),
+                    ter(tesSUCCESS));
+            env (
+                pay (gw, "alice", usdOffer), fee(f),
+                    ter(tesSUCCESS));
+            env (
+                offer ("alice", xrpOffer, usdOffer), fee(f),
+                    ter(tecINSUF_RESERVE_OFFER));
+
+            env.require (
+                balance ("alice", r - f),
```

Comment:
Let me see if I understand this, you're saying that *alice* pays `f` for the `trust` and then another `f` because the `offer` claimed a fee even though the reserve was insufficient?